### PR TITLE
Add drop cap project to spikes

### PR DESCRIPTION
**Pull request by Mecharyry**
This PR adds the DropCap project to spikes. Made up of a `DropCapView` and a `TypefaceFactory` using the android asset manager to create: 

drop_cap

### CODE REVIEW

File: drop_cap/README.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# DropCap
+This library can be used to create a DropCap in your app!
```
Comment (blundell) [inline drop_cap/README.md]: might be nice to add the definition of a dropcap :)

File: drop_cap/demo/src/main/java/com/mecharyry/dropcap/DropCapActivity.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+import android.app.Activity;
+import android.os.Bundle;
+import android.support.annotation.ColorInt;
+import android.view.View;
+import android.widget.AdapterView;
+import android.widget.ArrayAdapter;
+import android.widget.Button;
+import android.widget.Spinner;
+
+import com.mecharyry.drop_cap.R;
```
Comment (blundell) [inline drop_cap/demo/src/main/java/com/mecharyry/dropcap/DropCapActivity.java]: underscore in your package name, so gross :)

File: drop_cap/demo/src/main/java/com/mecharyry/dropcap/FontType.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package com.mecharyry.dropcap;
```
Comment (blundell) [inline drop_cap/demo/src/main/java/com/mecharyry/dropcap/FontType.java]: would having `demo` in the package name be less confusing (i.e. what is the library and what is the demo)

File: drop_cap/demo/src/main/java/com/mecharyry/dropcap/FontType.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import android.support.annotation.StringRes;
+
+import com.mecharyry.drop_cap.R;
+
+public enum FontType {
```
Comment (blundell) [inline drop_cap/demo/src/main/java/com/mecharyry/dropcap/FontType.java]: can this be package private?

File: drop_cap/demo/src/main/java/com/mecharyry/dropcap/SimpleSpinnerItemSelectedListener.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package com.mecharyry.dropcap;
+
+import android.view.View;
+import android.widget.AdapterView;
+import android.widget.Spinner;
+
+public class SimpleSpinnerItemSelectedListener implements Spinner.OnItemSelectedListener {
```
Comment (blundell) [inline drop_cap/demo/src/main/java/com/mecharyry/dropcap/SimpleSpinnerItemSelectedListener.java]: pacakage

File: drop_cap/demo/src/main/java/com/mecharyry/dropcap/TextColorDialogDisplayer.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
+package com.mecharyry.dropcap;
+
+import android.app.FragmentManager;
+import android.app.FragmentTransaction;
+import android.content.res.Resources;
+
+import com.mecharyry.drop_cap.R;
+
+public class TextColorDialogDisplayer {
```
Comment (blundell) [inline drop_cap/demo/src/main/java/com/mecharyry/dropcap/TextColorDialogDisplayer.java]: won't comment any more `package` :)

File: drop_cap/demo/src/main/java/com/mecharyry/dropcap/TextColorDialogDisplayer.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        this.textColorChangedListener = onTextColorChangeListener;
+    }
+
+    public void showTextColorDialog(int previousTextColor) {
+        String textColorFragmentTag = resources.getString(R.string.fragment_tag_text_color);
+        TextColorDialogFragment textColorFragment = (TextColorDialogFragment) fragmentManager.findFragmentByTag(textColorFragmentTag);
+
+        FragmentTransaction transaction = fragmentManager.beginTransaction();
+
+        if (textColorFragment == null) {
```
Comment (blundell) [inline drop_cap/demo/src/main/java/com/mecharyry/dropcap/TextColorDialogDisplayer.java]: 👍 preferring the positive

File: drop_cap/demo/src/main/res/layout/activity_drop_cap.xml [hunk 8]
```diff
@@ -0,0 +1,91 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+  android:layout_width="match_parent"
+  android:layout_height="match_parent"
+  android:orientation="vertical">
+
+  <FrameLayout
```
Comment (blundell) [inline drop_cap/demo/src/main/res/layout/activity_drop_cap.xml]: this is just a divider? Is FrameLayout the lightest thing you can use? maybe `Space` or `View`

File: drop_cap/demo/src/main/res/layout/activity_drop_cap.xml [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      android:id="@+id/drop_cap_font_spinner"
+      android:layout_width="0dp"
+      android:layout_height="wrap_content"
+      android:layout_weight="1" />
+
+    <Spinner
+      android:id="@+id/copy_font_spinner"
+      android:layout_width="0dp"
+      android:layout_height="wrap_content"
+      android:layout_weight="1" />
```
Comment (blundell) [inline drop_cap/demo/src/main/res/layout/activity_drop_cap.xml]: hey hey , better to have weights as a percent of 100

File: drop_cap/library/src/main/AndroidManifest.xml [hunk 10]
```diff
@@ -0,0 +1,2 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+  package="com.mecharyry.dropcap" />
```
Comment (blundell) [inline drop_cap/library/src/main/AndroidManifest.xml]: 💯  libs don't need most attrs

File: drop_cap/library/src/main/java/com/mecharyry/dropcap/DropCapView.java [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+    private int lineSpacingExtra;
+    private int numberOfLinesToSpan;
+    private int dropCapWidth;
+    private int dropCapLineHeight;
+    private float dropCapBaseline;
+
+    private boolean shouldDisplayDropCap;
+    private boolean hasPaintChanged;
+
+    public DropCapView(Context context, AttributeSet attrs) {
```
Comment (blundell) [inline drop_cap/library/src/main/java/com/mecharyry/dropcap/DropCapView.java]: what about the 3 arg constructor that ppl use in newer Android

File: drop_cap/library/src/main/java/com/mecharyry/dropcap/DropCapView.java [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+        hasPaintChanged = true;
+        requestLayout();
+    }
+
+    public float getDropCapTextSize() {
+        return dropCapPaint.getTextSize();
+    }
+
+    public void setDropCapTextColor(@ColorInt int color) {
+        if (color == dropCapPaint.getColor()) {
```
Comment (blundell) [inline drop_cap/library/src/main/java/com/mecharyry/dropcap/DropCapView.java]: love these quick returns, so much easier to think about

File: drop_cap/library/src/main/java/com/mecharyry/dropcap/DropCapView.java [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+    }
+
+    public void setText(String text) {
+        if (enoughTextForDropCap(text)) {
+            dropCapText = String.valueOf(text.charAt(0));
+            copyText = String.valueOf(text.subSequence(1, text.length()));
+            shouldDisplayDropCap = true;
+        } else {
+            dropCapText = String.valueOf('\0');
+            copyText = text == null ? "" : text;
```
Comment (blundell) [inline drop_cap/library/src/main/java/com/mecharyry/dropcap/DropCapView.java]: thats very gross, maybe adding parenthesis would help?

File: drop_cap/library/src/main/java/com/mecharyry/dropcap/TypefaceFactory.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+import android.graphics.Typeface;
+import android.util.AttributeSet;
+
+import java.lang.ref.SoftReference;
+import java.util.HashMap;
+import java.util.Map;
+
+class TypefaceFactory {
+
+    private AssetManager assetManager;
```
Comment (blundell) [inline drop_cap/library/src/main/java/com/mecharyry/dropcap/TypefaceFactory.java]: param ordering

**Comment by Mecharyry** [comment]
Thanks for the comments @blundell I'll try and address these comments ASAP 😄

File: drop_cap/README.md (hunk 1 above)
Comment (Mecharyry) [inline drop_cap/README.md]: fixed in afb884a

File: drop_cap/demo/src/main/java/com/mecharyry/dropcap/DropCapActivity.java (hunk 2 above)
Comment (Mecharyry) [inline drop_cap/demo/src/main/java/com/mecharyry/dropcap/DropCapActivity.java]: fixed in 2142075

File: drop_cap/demo/src/main/java/com/mecharyry/dropcap/FontType.java (hunk 3 above)
Comment (Mecharyry) [inline drop_cap/demo/src/main/java/com/mecharyry/dropcap/FontType.java]: fixed in 2142075

File: drop_cap/demo/src/main/java/com/mecharyry/dropcap/FontType.java (hunk 4 above)
Comment (Mecharyry) [inline drop_cap/demo/src/main/java/com/mecharyry/dropcap/FontType.java]: Fix packages cdbab77

File: drop_cap/demo/src/main/java/com/mecharyry/dropcap/SimpleSpinnerItemSelectedListener.java (hunk 5 above)
Comment (Mecharyry) [inline drop_cap/demo/src/main/java/com/mecharyry/dropcap/SimpleSpinnerItemSelectedListener.java]: Fix packages cdbab77

File: drop_cap/demo/src/main/java/com/mecharyry/dropcap/TextColorDialogDisplayer.java (hunk 6 above)
Comment (Mecharyry) [inline drop_cap/demo/src/main/java/com/mecharyry/dropcap/TextColorDialogDisplayer.java]: Fix packages cdbab77

File: drop_cap/demo/src/main/res/layout/activity_drop_cap.xml (hunk 8 above)
Comment (Mecharyry) [inline drop_cap/demo/src/main/res/layout/activity_drop_cap.xml]: fixed in d6e0179

File: drop_cap/demo/src/main/res/layout/activity_drop_cap.xml (hunk 9 above)
Comment (Mecharyry) [inline drop_cap/demo/src/main/res/layout/activity_drop_cap.xml]: Tots better :+1: 15a08be & 3abfcbd

File: drop_cap/library/src/main/java/com/mecharyry/dropcap/DropCapView.java (hunk 11 above)
Comment (Mecharyry) [inline drop_cap/library/src/main/java/com/mecharyry/dropcap/DropCapView.java]: Fixed in 	e2934bc

File: drop_cap/library/src/main/java/com/mecharyry/dropcap/DropCapView.java (hunk 13 above)
Comment (Mecharyry) [inline drop_cap/library/src/main/java/com/mecharyry/dropcap/DropCapView.java]: Totally gross :+1: 3cb56ae

File: drop_cap/library/src/main/java/com/mecharyry/dropcap/TypefaceFactory.java (hunk 14 above)
Comment (Mecharyry) [inline drop_cap/library/src/main/java/com/mecharyry/dropcap/TypefaceFactory.java]: and unused apparently :+1: 5cd6b6c

**Comment by blundell** [comment]
nice 👍

**Pull request merged**